[DPE-5527] Add Cryptography documentation (#50)
canonical/spark-k8s-bundle: releases/3.4/yaml/docs/explanation/e-security.md

# Cryptography

> This document describes cryptography used by Charmed Spark K8s.

Reviewer: Why is the file called security if we specifically talk about cryptography here? 
Is it supposed to be a security overview page for general security hardening or a dedicated page about cryptography?

## Authentication
In the Charmed Spark solution, authentication mechanism exists at the following places:

1. Connection with Zookeeper
2. Connection with PostgreSQL
3. Connection with S3 
4. Connection with Azure Storage
5. Kyuubi Client <> Kyuubi Server connection
6. Connection with Kubernetes Substrate
7. Connection between Spark Driver and Executors
8. Spark History Server Web Interface
 
### Connection with Zookeeper
Kyuubi Server authenticates to Zookeeper using hash digest-mechanism using digested MD5 hashes of username and password. Username and passwords are exchanged using databag relations between Kyuubi and ZooKeeper. These credentials are stored unencrypted in configuration files at /opt/kyuubi/conf inside the Kyuubi workload container.
Clients need to connect to Zookeeper for service discovery and retrieving information about the list of healthy Kyuubi Servers. The connection to ZooKeeper for service discovery is currently unauthenticated, exposing the list of healthy Kyuubi servers stored in a dedicated ZooKeeper node publicly readable by everyone.

### Connection with PostgreSQL

> Kyuubi Server authenticates to Postgres using username and password that are exchanged using relation data when the two applications are integrated. These credentials are stored by Kyuubi in configuration files at /opt/kyuubi/conf inside the workload container. More information can be found [here](https://kyuubi.readthedocs.io/en/master/security/jdbc.html).

Reviewer: ```suggestion
Kyuubi Server authenticates to Postgres using username and password that are exchanged using relation data when the two applications are integrated. These credentials are stored by Kyuubi in configuration files in the directory `/opt/kyuubi/conf` inside the workload container. For more information, see the [Kyuubi official documentation](https://kyuubi.readthedocs.io/en/master/security/jdbc.html).
```

### Connection to Object Storages

> Authentication is required for Spark Jobs, Spark History Server and Kyuubi to connect to either S3 or Azure object storages. Credentials are stored in peer-relation data for S3 Integrator and in Juju secrets for the Azure Object storage, and they are communicated to other charms (Spark History Server, Integration Hub and Kyuubi) via relation databag. Integration Hub stores the credentials into Kubernetes Secrets to be made available for Spark Jobs. Driver and executors store configuration files (where credential information is stored) unencrypted at /etc/spark8t/conf. Kyuubi stores credentials in unencrypted configuration files at /opt/spark/conf, to be used to configure Spark Engine, and /opt/kyuubi/conf, to be used to configure the Kyuubi server. Spark History Server stores credentials in unencrypted configuration files at /etc/spark/conf.

Reviewer: ```suggestion
Authentication is required for Spark Jobs, Spark History Server, and Kyuubi to connect to either S3 or Azure object storages. Credentials are stored in peer-relation data for S3 Integrator and in Juju secrets for the Azure Object storage, and they are communicated to other charms (Spark History Server, Integration Hub and Kyuubi) via relation databag. Integration Hub stores the credentials into Kubernetes Secrets to be made available for Spark Jobs. Driver and executors store configuration files (where credential information is stored) unencrypted in `/etc/spark8t/conf`. Kyuubi stores credentials in unencrypted configuration files in `/opt/spark/conf`, to be used to configure Spark Engine, and `/opt/kyuubi/conf`, to be used to configure the Kyuubi server. Spark History Server stores credentials in unencrypted configuration files in `/etc/spark/conf`.
```

 
### Connection between Spark Driver and Executors

> Spark internally has a capability to use a shared-secret based RPC authentication mechanism for connection between driver and executors to establish a secure channel. The authentication is not required , but can be enabled through dedicated [Spark properties](https://spark.apache.org/docs/latest/security.html#encryption).

Reviewer: ```suggestion
Spark internally has a capability to use a shared secret based RPC authentication mechanism for connection between driver and executors to establish a secure channel. The authentication is not required, but can be enabled through dedicated [Spark properties](https://spark.apache.org/docs/latest/security.html#encryption).
```

### Connection with Kubernetes Substrate
The connection between Charmed Spark and Kubernetes Substrate can be authenticated using various mechanisms like certificate based authentication, password based authentication, etc. which depends on how the underlying Kubernetes substrate is configured. 

### Kyuubi Client <> Kyuubi Server Connection

> The Kyuubi Client (aka JDBC Client) can connect with the Kyuubi Server endpoint (aka JDBC endpoint) using a pair of username and password provided in the query. Kyuubi charm internally implements this authentication by storing the list of users and their passwords stored as plaintext in a PostgreSQL database . More information about the Kyuubi client can be found [here](https://kyuubi.readthedocs.io/en/master/client/index.html).

Reviewer: ```suggestion
The Kyuubi Client (aka JDBC Client) can connect with the Kyuubi Server endpoint (aka JDBC endpoint) using a pair of a username and password provided in the query. Kyuubi charm internally implements this authentication by storing the list of users and their passwords stored as plaintext in a PostgreSQL database. For more information about the Kyuubi client, see the [Kyuubi documentation](https://kyuubi.readthedocs.io/en/master/client/index.html).
```

### Spark History Server Web Interface

> The Spark History Server provides authentication integrated using the Canonical Identity Platform in combination with a custom design [servlet filter](https://code.launchpad.net/~data-platform/soss/+source/charmed-spark-servlet-filters/+git/charmed-spark-servlet-filters).This feature can be enabled by following the step in the following [document](https://charmhub.io/spark-k8s-bundle/docs/h-history-server-authorization). 

Reviewer: ```suggestion
The Spark History Server provides authentication integrated using the Canonical Identity Platform in combination with a custom design [servlet filter](https://code.launchpad.net/~data-platform/soss/+source/charmed-spark-servlet-filters/+git/charmed-spark-servlet-filters). This feature can be enabled by following the step in the [History server authorisation How to guide](https://charmhub.io/spark-k8s-bundle/docs/h-history-server-authorization). 
```
